skada/_subspace.py

```diff
@@ -366,3 +373,265 @@ def TransferComponentAnalysis(
+            X_ = K @ self.A_
+        else:
+            Ks = pairwise_kernels(X, self.X_source_, metric=self.kernel)
+            Kt = pairwise_kernels(X, self.X_target_, metric=self.kernel)
+            K = np.concatenate((Ks, Kt), axis=1)
+            X_ = K @ self.A_
+        return X_
+
+    def _get_mmd_matrix(self, ns, nt, sample_domain):
+        Mss = (1 / (ns**2)) * np.ones((ns, ns))
+        Mtt = (1 / (nt**2)) * np.ones((nt, nt))
+        Mst = -(1 / (ns * nt)) * np.ones((ns, nt))
+        M = np.block([[Mss, Mst], [Mst.T, Mtt]])
+        return M
+
+    def _get_kernel_matrix(self, X_source, X_target):
+        Kss = pairwise_kernels(X_source, metric=self.kernel)
+        Ktt = pairwise_kernels(X_target, metric=self.kernel)
+        Kst = pairwise_kernels(X_source, X_target, metric=self.kernel)
+        K = np.block([[Kss, Kst], [Kst.T, Ktt]])
+        return K
+
+    def fit(self, X, y=None, sample_domain=None, **kwargs):
+        """Fit adaptation parameters.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The source data.
+        y : array-like, shape (n_samples,)
+            The source labels.
+        sample_domain : array-like, shape (n_samples,)
+            The domain labels (same as sample_domain).
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        X, sample_domain = check_X_domain(
+            X,
+            sample_domain,
+            allow_multi_source=True,
+            allow_multi_target=True,
+        )
+        X_source, X_target = source_target_split(X, sample_domain=sample_domain)
+
+        if self.n_components is None:
+            n_components = min(min(X_source.shape), min(X_target.shape))
+        else:
+            n_components = self.n_components
+        self.random_state_ = check_random_state(self.random_state)
+        self.X_source_ = X_source
+        self.X_target_ = X_target
+
+        n = X.shape[0]
+        source_mask = extract_source_indices(sample_domain)
+
+        H = np.identity(n) - 1 / n * np.ones((n, n))
+        K = self._get_kernel_matrix(X_source, X_target)
+        M = self._get_mmd_matrix(X_source.shape[0], X_target.shape[0], sample_domain)
+        M /= np.linalg.norm(M)
+        G = np.identity(n)
+
+        # minimization of the objective function
+        # \min_{A} tr(A^T K M K A) + tradeoff * (||A_s||_{2, 1} + ||A_t||_F^2)
+        # s.t. A^T K H K^T A = I
+        EPS_eigval = 1e-12
+        for i in range(self.max_iter):
+            # update A
+            B = K @ M @ K.T + self.tradeoff * G
+            B = B + EPS_eigval * np.identity(n)
+            C = K @ H @ K.T
+            C = C + EPS_eigval * np.identity(n)
+            solution = scipy.linalg.solve(B, C)
+            phi, A = scipy.linalg.eigh(solution)
+            phi = phi + EPS_eigval
+            indices = np.argsort(-np.abs(phi))[:n_components]
```

Comment:
`B` and `C` are symmetric (semi-) positive definite. So eigenvalues `phi` must be positive. So you have to remove `np.abs`, otherwise there is a problem somewhere.

You put a minus, so get the greatest eigenvalues. If I remember correctly, the algorithm requires the lowest one.